### PR TITLE
ENH: More elegant tripleg generation

### PR DESCRIPTION
**Pull request by henrymartin1**
Here is a first draft of a pull request that proposes a new and simpler tripleg generation.

### CODE REVIEW

**Comment by henrymartin1** [comment]
fixes #27

**Comment by henrymartin1** [comment]
I had a problem with the last commit message in `d1a7344` but this commit finishes the code however due to shifting the tripleg definition by 1 point (tripleg starts with the first point after a staypoint instead of the shared last point of the staypoint) the tests have to updated

**Comment by svenruf** [comment]
I assume that you will also update the tests as you assigned yourself, right? If there's anything to do for me, let me know.

File: trackintel/preprocessing/positionfixes.py [hunk 1]
```diff
@@ -261,26 +260,30 @@ def generate_triplegs(positionfixes_in, spts=None, method='between_staypoints',
...
+        #  `cond_all = cond_all & pd.isna(pfs['staypoint_id'])`
         tpls_starts = np.where(cond_all)[0]
         tpls_lengths = np.diff(tpls_starts)
-        cond_to_remove = np.take(tpls_starts,np.where(tpls_lengths<3)[0])
-        for i in cond_to_remove:
-            cond_all[i] = False
-            pfs.loc[i,'tripleg_id'] = -1
+        cond_to_remove = np.take(tpls_starts, np.where(tpls_lengths < 2)[0])  # a valid linestring needs 2 points
+        cond_all.iloc[cond_to_remove] = False
+        pfs.iloc[cond_to_remove].loc['tripleg_id'] = -1
```
Comment (svenruf) [inline trackintel/preprocessing/positionfixes.py:286]: This causes a warning when I execute it, as it is trying to set the -1 on a copy of pfs

File: trackintel/preprocessing/positionfixes.py [hunk 2]
```diff
@@ -204,52 +204,51 @@ def generate_triplegs(positionfixes_in, spts=None, method='between_staypoints',
...
+                # step 2
+                # Identify first positionfix after a staypoint
+                # find index of closest positionfix with equal or greater timestamp.
                 tracked_at_sorted = pfs_u['tracked_at'].sort_values()
                 insert_position_psf_u = tracked_at_sorted.searchsorted(spts_u['finished_at'])
                 insert_index_psf = tracked_at_sorted.iloc[insert_position_psf_u].index
 
                 # store the insert insert_position_spts_u in an array
-                insert_index_psf_list.extend(list(tracked_at_sorted.iloc[insert_position_psf_u].index))
+                insert_index_psf_list.extend(list(tracked_at_sorted.iloc[insert_index_psf].index))
```
Comment (svenruf) [inline trackintel/preprocessing/positionfixes.py:241]: This causes a out of bound error when testing it with the geolife_long. removing the .iloc solves the error

**Comment by svenruf** [comment]
@hong2223 as discussed I had a look and ran the tripleg generation. At the moment there are two things that need to have closer look, I commented directly into Henrys commit. I did not want to push the changes directly, as Henry did change this from my code and I don't know if it has a special reason and needs another solution.

File: trackintel/preprocessing/positionfixes.py (hunk 1 above)
Comment (hong2223) [reply trackintel/preprocessing/positionfixes.py:286]: I changed this now to `pfs.loc[pfs.index.isin(cond_to_remove), "tripleg_id"] = -1`

File: trackintel/preprocessing/positionfixes.py (hunk 2 above)
Comment (hong2223) [reply trackintel/preprocessing/positionfixes.py:241]: I think this is actually `insert_index_ls.extend(list(insert_index_psf))`

**Comment by hong2223** [comment]
> @hong2223 as discussed I had a look and ran the tripleg generation. At the moment there are two things that need to have closer look, I commented directly into Henrys commit. I did not want to push the changes directly, as Henry did change this from my code and I don't know if it has a special reason and needs another solution.

Thanks for the check! I had a look at the function and completed its functionality. 
I messed up with the commits, so I will close this one (again!) and open a new one - -..

**Pull request closed without merging**